packages/typespec-azure-resource-manager/src/private.decorators.ts

```diff
@@ -350,3 +350,18 @@ function hasProperty(program: Program, model: Model): boolean {
   if (model.baseModel) return hasProperty(program, model.baseModel);
   return false;
 }
+
+/**
+ * Please DO NOT USE in RestAPI specs.
+ * Internal decorator that deprecated direct usage of `x-ms-client-flatten` OpenAPI extension.
+ * It will programatically enabled/disable client flattening with @flattenProperty with autorest
+ * emitter flags to maintain compatibility in swagger.
+ */
+export function $conditionalClientFlatten(context: DecoratorContext, entity: ModelProperty) {
+  context.program.stateMap(ArmStateKeys.armConditionalClientFlatten).set(entity, true);
+}
+
+export function isConditionallyFlattened(program: Program, entity: ModelProperty): boolean {
```

Comment:
would naming this legacyFlattened be better?

Reply:
if legacy is likely referring to the legacy client? I'd like to point it out that the legacy client behavior is different from language to language, for example  Go doesn't support flatten before. while other languages like dotnet, js does.

Reply:
Per conversation with Laurent, this function is intended for `autorest`. We are done away with flattening by default. If folks wants to flatten, instead of relying this decorator, spec writer can use `@@flattenProperty(XX.properties)` in `client.tsp`. Let's keep the name as it is.